cpp/src/join/quadtree_point_in_polygon.cu

```diff
@@ -192,191 +108,102 @@ struct dispatch_quadtree_point_in_polygon {
-                                       0,
-                                       thrust::plus<uint32_t>());
-
-      // allocate memory for sub-pairs' quad_offset component
-      rmm::device_uvector<uint32_t> d_quad_offset(num_unit_pairs, stream);
-
-      cudaMemsetAsync(d_quad_offset.data(), 0, d_quad_offset.size() * sizeof(uint32_t), stream);
-
-      // scatter 0..num_original_pairs to d_quad_offset using d_unit_offsets as the scatter map
-      thrust::scatter(rmm::exec_policy(stream)->on(stream),
-                      thrust::make_counting_iterator(0),
-                      thrust::make_counting_iterator(0) + num_original_pairs,
-                      d_unit_offsets.begin(),
-                      d_quad_offset.begin());
-
-      // copy idx of orginal pairs to all sub-pairs
-      thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream),
-                             d_quad_offset.begin(),
-                             d_quad_offset.end(),
-                             d_quad_offset.begin(),
-                             thrust::maximum<int>());
-
-      // d_unit_offsets is no longer needed
-
-      return d_quad_offset;
-    }();
-
-    // allocate memory for the sub-pairs' other three components (poly_idx, quad_idx, quad_length)
-    rmm::device_uvector<uint32_t> d_pq_poly_idx(num_unit_pairs, stream);
-    rmm::device_uvector<uint32_t> d_pq_quad_idx(num_unit_pairs, stream);
-    rmm::device_uvector<uint32_t> d_quad_length(num_unit_pairs, stream);
-
-    // gather polygon idx and quadrant idx from original pairs into sub-pairs
-    thrust::gather(rmm::exec_policy(stream)->on(stream),
-                   d_quad_offset.begin(),
-                   d_quad_offset.end(),
-                   poly_indices.begin<uint32_t>(),
-                   d_pq_poly_idx.begin());
-
-    thrust::gather(rmm::exec_policy(stream)->on(stream),
-                   d_quad_offset.begin(),
-                   d_quad_offset.end(),
-                   quad_indices.begin<uint32_t>(),
-                   d_pq_quad_idx.begin());
-
-    // generate offsets of sub-pairs within the orginal pairs
-    thrust::exclusive_scan_by_key(rmm::exec_policy(stream)->on(stream),
-                                  d_quad_offset.begin(),
-                                  d_quad_offset.end(),
-                                  thrust::constant_iterator<int>(1),
-                                  d_quad_offset.begin());
-
-    // assemble components in input/output iterators; note d_quad_offset used in both input and
-    // output
-    auto quad_block_id_iter = make_zip_iterator(d_pq_quad_idx.begin(), d_quad_offset.begin());
-    auto offset_length_iter = make_zip_iterator(d_quad_offset.begin(), d_quad_length.begin());
+    auto quad_lengths        = quadtree.column(3);
+    auto quad_offsets        = quadtree.column(4);
+    auto poly_indices        = poly_quad_pairs.column(0);
+    auto quad_indices        = poly_quad_pairs.column(1);
```

Comment:
A thought, possibly for a different PR: we should probably encode the column indices somewhere (so you can access by name rather than number), or consider making a quadtree a data structure rather than a table -- I don't think anyone will pass in a table that is a quadtree from outside cuSpatial -- they will probably build it using cuSpatial, right?

Reply:
I agree, making it a data structure would make things simpler all the way up to Python.